gwcs/wcs.py

```diff
@@ -1716,42 +2105,440 @@ def to_fits_tab(self, bounding_box=None, bin_ext_name='WCS-TABLE',
+        input_axes = []
+        world_axes_idx = []
+        for ax in world_axes_group:
+            world_axes_idx.append(ax.axis)
+            input_axes.extend(ax.input_axes)
+        input_axes = sorted(set(input_axes))
+        n_inputs = len(input_axes)
+        n_outputs = len(world_axes_group)
+        world_axes_idx.sort()
+
+        # Create initial header and deal with non-degenerate axes
+        if hdr is None:
+            hdr = fits.Header()
+            hdr['NAXIS'] = n_inputs, 'number of array dimensions'
+            hdr['WCSAXES'] = n_outputs
+            hdr.insert('WCSAXES', ('WCSNAME', f'{self.output_frame.name}'), after=True)
+
+        else:
+            hdr['NAXIS'] += n_inputs
+            hdr['WCSAXES'] += n_outputs
+
+        # see what axes have been already populated in the header:
+        used_hdr_axes = []
+        for v in hdr['naxis*'].keys():
+            try:
+                used_hdr_axes.append(int(v.split('NAXIS')[1]) - 1)
+            except ValueError:
+                continue
+
+        degenerate_axis_start = max(
+            self.pixel_n_dim + 1,
+            max(used_hdr_axes) + 1 if used_hdr_axes else 1
+        )
+
+        # Deal with non-degenerate axes and add NAXISi to the header:
+        offset = hdr.index('NAXIS')
+
+        for iax in input_axes:
+            iiax = int(np.searchsorted(used_hdr_axes, iax))
+            hdr.insert(iiax + offset + 1, (f'NAXIS{iax + 1:d}', int(max(bounding_box[iiax])) + 1))
+
         # 1D grid coordinates:
         gcrds = []
         cdelt = []
-        for (xmin, xmax), s in zip(bounding_box, sampling):
+        bb = [bounding_box[k] for k in input_axes]
+        for (xmin, xmax), s in zip(bb, sampling):
             npix = max(2, 1 + int(np.ceil(abs((xmax - xmin) / s))))
             gcrds.append(np.linspace(xmin, xmax, npix))
             cdelt.append((npix - 1) / (xmax - xmin) if xmin != xmax else 1)
 
-        # n-dim coordinate arrays:
+        # In the forward transformation, select only inputs and outputs
+        # that we need given world_axes_group parameter:
+        bb_center = np.mean(bounding_box, axis=1)
+
+        fixi_dict = {
+            k: bb_center[k] for k in set(range(self.pixel_n_dim)).difference(input_axes)
+        }
+
+        transform = _fix_transform_inputs(self.forward_transform, fixi_dict)
+        transform = transform | Mapping(world_axes_idx,
+                                        n_inputs=self.forward_transform.n_outputs)
+
+        xyz = np.meshgrid(*gcrds[::-1], indexing='ij')[::-1]
+
+        shape = xyz[0].shape
+        xyz = [v.ravel() for v in xyz]
+
         coord = np.stack(
-            self(*np.meshgrid(*gcrds[::-1], indexing='ij')[::-1]),
+            transform(*xyz),
             axis=-1
         )
 
+        coord = coord.reshape(shape + (len(world_axes_group), ))
+
         # create header with WCS info:
-        hdr = fits.Header()
+        if hdr is None:
+            hdr = fits.Header()
```

Comment:
Would it be possible to break out the actual writing of the FITS header into a separate function.  It seems `_to_fits_tab()` is carrying a lot of water, being both a calling wrapper and header producer.  Or is that too much hassle given the number of parameters needing to be passed?

Reply:
I thought about this but it is too much hassle. The number of parameters is variable based on enabling SIP or not and number of axes and their separability. Maybe using a dictionary would be OK but that is what basically a header is.